### ISSUE TITLE
./include/libmesh/parallel_implementation.h:550:15: error: invalid conversion from ‘const void*’ to ‘void*’

### ISSUE THREAD

**Issue by wgm096350** [libMesh/libmesh#2243]
OpenMPI 1.6.5 have been installed on my PC.
I got the latest code via "git clone git://github.com/libMesh/libmesh.git ". My build process is:
**cd <libmesh dir>
mkdir build
cd ./build
../configure --disable-fortran**
with these commands, the makefile is generated. Then:
**make**
but this "make" command ends up with the following error:
**../src/base/dof_map.C:991:61:   required from here
./include/libmesh/parallel_implementation.h:550:15: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]
     (MPI_Pack (&mpi_n_vecs, 1,
      ~~~~~~~~~^~~~~~~~~~~~~~~~
                libMesh::Parallel::StandardType<unsigned int>(),
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                sendbuf->data(), sendsize, &pos, this->get()));
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~**

Has anyone encounter this problem before?

**Comment by wgm096350** [comment]
here is  the config.log file
[config.log](https://github.com/libMesh/libmesh/files/3596631/config.log)

**Comment by jwpeterson** [comment]
I think this is a MPI-2 vs. MPI-3 compatibility issue. There was some (I think) related discussion on #1827. What version of libmesh (git hash, etc.) are you using?

**Comment by roystgnr** [comment]
Yeah; compare [the old API](https://www.open-mpi.org/doc/v1.6/man3/MPI_Pack.3.php) vs [the new API](https://www.open-mpi.org/doc/v4.0/man3/MPI_Pack.3.php) for MPI_Pack - the lack of that const specifier for the input argument is what's getting you.

We're no longer supporting MPI implementations that old (note that the OpenMPI site calls it "ancient"; even 1.10 and 2.0 are just "retired" and unsupported at this point), but I do apologize for you seeing a confusing error message at compile time rather than a clear one at configure time.  Getting #1920 in shape to merge is still on my TODO list but keeps getting buried by more urgent things.

**Comment by jwpeterson** [comment]
On Ubuntu 18, there are newer OpenMPI implementations available. For example:
```
$ apt show openmpi-common
Package: openmpi-common
Version: 2.1.1-8
```
If you have administrator rights on this system, you could try simply updating OpenMPI and that should fix the problem... but presumably it is installed on your system for some reason so this might not be feasible.

**Comment by wgm096350** [comment]
> Yeah; compare [the old API](https://www.open-mpi.org/doc/v1.6/man3/MPI_Pack.3.php) vs [the new API](https://www.open-mpi.org/doc/v4.0/man3/MPI_Pack.3.php) for MPI_Pack - the lack of that const specifier for the input argument is what's getting you.
> 
> We're no longer supporting MPI implementations that old (note that the OpenMPI site calls it "ancient"; even 1.10 and 2.0 are just "retired" and unsupported by them at this point), but I do apologize for you seeing a confusing error message at compile time rather than a clear one at configure time. Getting #1920 in shape to merge is still on my TODO list but keeps getting buried by more urgent things.

thanks for your reply. I will try updating MPI. your commitment is greatly appreciated

**Comment by wgm096350** [comment]
> On Ubuntu 18, there are newer OpenMPI implementations available. For example:
> 
> ```
> $ apt show openmpi-common
> Package: openmpi-common
> Version: 2.1.1-8
> ```
> 
> If you have administrator rights on this system, you could try simply updating OpenMPI and that should fix the problem... but presumably it is installed on your system for some reason so this might not be feasible.

yeah. other users may rely on the "ancient" 1.6.5 version. is there a way for an new and an old version of MPI existing at the same time?

**Comment by roystgnr** [comment]
> is there a way for an new and an old version of MPI existing at the same time?

If you build your own, it's easy; just put them in different directories, and make users decide which they want and set LDFLAGS and such accordingly.

If you've already got one installed in a system library path (probably /usr/lib/x86_64-linux-gnu/ for you) then it's a mess; you need that path for other things your linker will want, and it's hard to juggle the new path in linker arguments without ever inadvertently pulling in the wrong version of the library.

**Comment by wgm096350** [comment]
> > is there a way for an new and an old version of MPI existing at the same time?
> 
> If you build your own, it's easy; just put them in different directories, and make users decide which they want and set LDFLAGS and such accordingly.
> 
> If you've already got one installed in a system library path (probably /usr/lib/x86_64-linux-gnu/ for you) then it's a mess; you need that path for other things your linker will want, and it's hard to juggle the new path in linker arguments without ever inadvertently pulling in the wrong version of the library.

currently. mpi1.6.5 is installed in /usr/local/bin by someone else. I newly installed version 2.1.6 in /home/<my username>/openmpi216. Then I added "export PATH=/home/<my username>/openmpi216:$PATH" to ~/.bashrc. After this,I repeat the configure,make,make install process. After installation, cd to the /<installation di>/bin, run "./libmesh-config --cxx", the termimal outputs "mpicxx", run "which mpicxx", outputs "/home/<my username>/openmpi216/bin/mpicxx". Is this solution OK?

**Comment by roystgnr** [comment]
nCalls can change when we change the code to profile more or less or to run more efficiently; that's fine.

You can see in the output of configure whether or not libMesh liked and is building against your PETSc install.  But at this point, you could just try ```mpirun -np 4 ./example-opt -ksp_type cg``` and make sure n_local_elem() is about one fourth of n_elem(); if so then it looks like you've got PETSc working.

**Comment by wgm096350** [comment]
> nCalls can change when we change the code to profile more or less or to run more efficiently; that's fine.
> 
> You can see in the output of configure whether or not libMesh liked and is building against your PETSc install. But at this point, you could just try `mpirun -np 4 ./example-opt -ksp_type cg` and make sure n_local_elem() is about one fourth of n_elem(); if so then it looks like you've got PETSc working.

Thanks.
But currently I don't know how to change local element. I only know i should modify this snippet in systems_of_equations_ex5.C to change the mesh but not how:
`  Mesh mesh(init.comm(), dim);
  MeshTools::Generation::build_square (mesh,
                                       50, 10,
                                       0., 1.,
                                       0., 0.2,
                                       QUAD9);`
Are there any manuals or tutorials about using libmesh?

**Comment by wgm096350** [comment]
> > nCalls can change when we change the code to profile more or less or to run more efficiently; that's fine.
> > You can see in the output of configure whether or not libMesh liked and is building against your PETSc install. But at this point, you could just try `mpirun -np 4 ./example-opt -ksp_type cg` and make sure n_local_elem() is about one fourth of n_elem(); if so then it looks like you've got PETSc working.
> 
> Thanks.
> But currently I don't know how to change local element. I only know i should modify this snippet in systems_of_equations_ex5.C to change the mesh but not how:
> ` Mesh mesh(init.comm(), dim); MeshTools::Generation::build_square (mesh, 50, 10, 0., 1., 0., 0.2, QUAD9);`
> Are there any manuals or tutorials about using libmesh?

Oh. n_local_elem() is set automatically. This is the result when I run `mpirun -np 4 ./example-opt -ksp_type cg`. Is this enough to demonstrate that PETSC and MPI working?
[mpirun_res.log](https://github.com/libMesh/libmesh/files/3612205/mpirun_res.log)

**Comment by roystgnr** [comment]
Yup, you're running with PETSc solvers in parallel just fine there.

**Comment by BalticPinguin** [comment]
... getting back to the original point, maybe instead of the checks for ``MPI_VERSION`` etc. done in #1920, we can just check the API, i.e. if it is const or non-const.

To be honest, #1920 escaped my attention with the time, I hope I can get back to this soon.
